packages/react-instantsearch/src/components/StarRating.js

```diff
@@ -128,39 +127,47 @@ class StarRating extends Component {
   }
 
   render() {
-    const {
-      translate,
-      refine,
-      min,
-      max,
-      count,
-      createURL,
-      canRefine,
-    } = this.props;
-    const items = [];
-    for (let i = max; i >= min; i--) {
-      const hasCount = !isEmpty(count.filter(item => Number(item.value) === i));
-      const lastSelectableItem = count.reduce(
-        (acc, item) =>
-          item.value < acc.value || (!acc.value && hasCount) ? item : acc,
-        {}
-      );
-      const itemCount = count.reduce(
-        (acc, item) => (item.value >= i && hasCount ? acc + item.count : acc),
-        0
-      );
-      items.push(
-        this.buildItem({
-          lowerBound: i,
-          max,
-          refine,
-          count: itemCount,
-          translate,
-          createURL,
-          isLastSelectableItem: i === Number(lastSelectableItem.value),
-        })
+    const { min, max, translate, count, createURL, canRefine } = this.props;
+
+    // min & max are always set when there is a results, otherwise it means
+    // that we don't want to render anything since we don't have any values.
+    const limitMin = min !== undefined && min >= 0 ? min : 0;
+    const limitMax = max !== undefined && max >= 0 ? max : -1;
+    const inclusiveLength = limitMax - limitMin + 1;
+    const safeInclusiveLength = Math.max(inclusiveLength, 0);
+
+    const values = count
+      .map(item => ({ ...item, value: parseFloat(item.value) }))
+      .filter(item => item.value >= limitMin && item.value <= limitMax);
+
+    const range = new Array(safeInclusiveLength)
+      .fill(null)
+      .map((_, index) => {
+        const element = values.find(item => item.value === limitMax - index);
+        const placeholder = { value: limitMax - index, count: 0, total: 0 };
+
+        return element || placeholder;
+      })
+      .reduce(
+        (acc, item, index) =>
+          acc.concat({
+            ...item,
+            total: index === 0 ? item.count : acc[index - 1].total + item.count,
+          }),
+        []
       );
-    }
+
+    const items = range.map((item, index) =>
+      this.buildItem({
+        lowerBound: item.value,
+        count: item.total,
+        isLastSelectableItem: range.length - 1 === index,
+        max: limitMax,
+        translate,
+        createURL,
```

Comment:
why is `refine` not in this object anymore?

Reply:
Take a look at the signature of `buildItem` refine is not used. We access the function through the `props`.

https://github.com/algolia/react-instantsearch/blob/4064e0cac7dc96a23565a7f33d11b4de0e253434/packages/react-instantsearch/src/components/StarRating.js#L52-L59